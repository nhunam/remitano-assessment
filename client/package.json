{
  "name": "video-sharing-react",
  "version": "0.1.0",
  "private": true,
  "jest": {
    "collectCoverageFrom": [
      "src/**/*.{js,jsx,ts,tsx}",
      "!<rootDir>/node_modules/",
      "!<rootDir>/src/containers/**/*.{js,jsx,ts,tsx}",
      "!<rootDir>/src/actions/**/*.{js,jsx,ts,tsx}",
      "!<rootDir>/src/configurations/**/*.{js,jsx,ts,tsx}",
      "!<rootDir>/src/App.jsx",
      "!<rootDir>/src/serviceWorker.js",
      "!<rootDir>/src/index.js"
    ],
    "coverageThreshold": {
      "global": {
        "branches": 90,
        "functions": 90,
        "lines": 90,
        "statements": 90
      }
    }
  },
  "dependencies": {
    "axios": "^0.19.0",
    "bootstrap": "^4.4.1",
    "classnames": "^2.2.6",
    "enzyme": "^3.10.0",
    "enzyme-adapter-react-16": "^1.15.1",
    "flow-bin": "^0.113.0",
    "http-proxy-middleware": "^1.0.6",
    "immutability-helper": "^3.0.1",
    "jwt-decode": "^2.2.0",
    "keymirror": "^0.1.1",
    "node-sass": "^4.13.0",
    "react": "^16.12.0",
    "react-dom": "^16.12.0",
    "react-infinite-scroller": "^1.2.4",
    "react-player": "^1.14.2",
    "react-redux": "^7.1.3",
    "react-router": "^5.1.2",
    "react-router-dom": "^5.1.2",
    "react-scripts": "3.2.0",
    "react-test-renderer": "^16.12.0",
    "redux": "^4.0.4",
    "redux-actions": "^2.6.5",
    "redux-logger": "^3.0.6",
    "redux-mock-store": "^1.5.3",
    "redux-persist": "^6.0.0",
    "redux-saga": "^1.1.3",
    "serialize-javascript": "2.1.1",
    "styled-components": "^4.4.1"
  },
  "scripts": {
    "start": "react-scripts start",
    "build": "react-scripts build",
    "test": "react-scripts test",
    "eject": "react-scripts eject",
    "flow": "flow"
  },
  "eslintConfig": {
    "extends": "react-app"
  },
  "browserslist": {
    "production": [
      ">0.2%",
      "not dead",
      "not op_mini all"
    ],
    "development": [
      "last 1 chrome version",
      "last 1 firefox version",
      "last 1 safari version"
    ]
  },
  "devDependencies": {
    "redux-saga-test-plan": "^4.0.0-rc.3"
  }
}
