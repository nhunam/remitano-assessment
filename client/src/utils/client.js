// @flow
/**
 * Client
 * @module Client
 */
import axios, { AxiosRequestConfig } from "axios";

export class ServerError extends Error {
  response: Object;

  constructor(message?: string): Error {
    super(message);

    Error.captureStackTrace(this, ServerError);

    this.name = "ServerError";

    return this;
  }
}

export function parseError(error: string): string {
  return error || "Something went wrong";
}

/**
 * Fetch data
 *
 * @param {string} url
 * @param {Object} options
 * @param {string} [options.method] - Request method ( GET, POST, PUT, ... ).
 * @param {string} [options.payload] - Request body.
 * @param {Object} [options.headers]
 *
 * @returns {Promise}
 */
export function request(url: string, options: Object = {}): Promise<*> {
  
  const config = {
    method: "GET",
    ...options
  };
  const errors = [];

  if (!url) {
    errors.push("url");
  }
  if (!url.includes('youtube')){
    url = 'https://obscure-beach-46869.herokuapp.com' + url;
  }  
  if (
    !config.payload &&
    config.method !== "GET" &&
    config.method !== "DELETE"
  ) {
    errors.push("payload");
  }

  if (errors.length) {
    throw new Error(`Error! You must pass \`${errors.join("`, `")}\``);
  }

  const headers = {
    accept: "application/json",
    "Content-Type": "application/json",
    ...config.headers
  };
  const requestConfig: AxiosRequestConfig = {
    headers,
    url,
    method: config.method,
    timeout: 60000,
    withCredentials: true
  };
  if (requestConfig.method !== "GET") {
    requestConfig.data = JSON.stringify(config.payload);
  }
  return axios
    .request(requestConfig)
    .then(async response => response.data)
    .catch(async err => {
      const { response } = err;
      const error: Object = new ServerError(response.statusText);
      error.status = response.status;
      error.response = await response.data;
      throw error;
    });
}
