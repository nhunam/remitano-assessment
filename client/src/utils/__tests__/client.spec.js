import * as axios from "axios";

import { ServerError, parseError, request } from "utils/client";

jest.mock("axios");

describe("client", () => {
  beforeEach(() => {});

  describe("request", () => {
    it("should fail without param", () => {
      expect(() => request()).toThrow("Error! You must pass `url`");
    });

    it("should fail to a POST without payload", () => {
      expect(() =>
        request("http://example.com/token", { method: "POST" })
      ).toThrow("Error! You must pass `payload`");
    });

    it("should execute a GET successfully", done => {
      axios.request.mockImplementation(() =>
        Promise.resolve({
          data: JSON.stringify({ hello: "world" }),
          ...global.fetchInit
        })
      );
      request("http://example.com/token").then(data => {
        expect(data).toMatchSnapshot();
        done();
      });
    });

    it("should execute a POST successfully", done => {
      axios.request.mockImplementation(() =>
        Promise.resolve({
          data: null,
          status: 201,
          statusText: "Created",
          ...global.fetchInit
        })
      );
      request("http://example.com/token", {
        method: "POST",
        payload: { a: 1 }
      }).then(data => {
        expect(data).toMatchSnapshot();
        done();
      });
    });

    it("should reject for a  bad request", done => {
      const err = new Error();
      err.response = {
        data: { error: "FAILED" },
        status: 400,
        statusText: "Invalid",
        ...global.fetchInit
      };
      axios.request.mockImplementation(() => Promise.reject(err));
      request("http://example.com/token").catch(error => {
        expect(error.response).toEqual({ error: "FAILED" });
        expect(error.status).toBe(400);
        done();
      });
    });
  });

  describe("ServerError", () => {
    it("should create a new instance", () => {
      const error = new ServerError("error");
      expect(error.name).toBe("ServerError");
    });
  });

  describe("parseError", () => {
    it("should handle messages", () => {
      expect(parseError("Failed")).toBe("Failed");
      expect(parseError()).toBe("Something went wrong");
    });
  });
});
