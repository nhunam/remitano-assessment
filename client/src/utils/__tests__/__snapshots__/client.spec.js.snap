// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client request should execute a GET successfully 1`] = `"{\\"hello\\":\\"world\\"}"`;

exports[`client request should execute a POST successfully 1`] = `null`;
